python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -427,23 +427,31 @@ def filter_types(records: Iterable[CollectionRecord[K]]) -> Iterator[CollectionR
         cache: CollectionRecordCache | None = None
         result: list[CollectionRecord[K]] = []
         done_keys: set[K] = set()
+        explicit_names = list(wildcard.strings)
         if wildcard.patterns:
-            if wildcard.strings:
+            if explicit_names or flatten_chains:
                 # To be efficient in case both patterns and strings are
                 # specified we want to have caching enabled for at least the
-                # duration of this call.
+                # duration of this call. Chains flattening can also produce
+                # additional names to look for.
                 cache = self._caching_context.collection_records or CollectionRecordCache()
             all_records = self._fetch_all(cache)
             for record in filter_types(all_records):
                 if record.key not in done_keys:
                     if any(p.fullmatch(record.name) for p in wildcard.patterns):
                         result.append(record)
                         done_keys.add(record.key)
-
-        if wildcard.strings:
+            if flatten_chains:
+                # If flattening include children names of all matching chains.
+                for record in all_records:
+                    if isinstance(record, ChainedCollectionRecord):
+                        if any(p.fullmatch(record.name) for p in wildcard.patterns):
+                            explicit_names.extend(record.children)
```

Comment:
Does this new logic need to be recursive, or are we already in something that's recursing here?